Fix JsonParser throwing inconsistent exceptions or errors

Depending on the D version used, JsonParser will throw a different exception or error. In order to make it consistent it should throw always a `JsonParserException` so a program using the module can catch the exception whatever is the D version currently in use.
 
```D
unittest
{
    static immutable istring json =
    `{
        "products": {
        }}
    }`;

    auto p = new JsonParser!(char)(json);
    assert(p);
    assert(p.type == p.Token.BeginObject);
    assert(p.next);
    assert(p.type == p.Token.Name);
    assert(p.value == "products", p.value);
    assert(p.next);
    assert(p.value == "", p.value);
    assert(p.next);
    assert(p.type == p.Token.EndObject);
    assert(p.next);
    assert(p.type == p.Token.EndObject);

    try
    {
        assert(p.next);
    }
    catch (Exception e)
    {
        // catch in D1
    }
    catch
    {
        // catch in D2
    }
}
```

Thanks for the clarification. I found it a bit confusing TBH reading the message from the stack trace.

<details>
<summary>This is the stack trace from D2 if the `catch` is removed from the unit test</summary>

```
core.exception.RangeError@./src/ocean/util/container/more/Stack.d(169): Range violation
----------------
./src/ocean/util/container/more/Stack.d:295 pure nothrow @safe ocean.text.json.JsonParser.JsonParser!(char, false).JsonParser.State ocean.util.container.more.Stack.__T5StackTE5ocean4text4json10JsonParser22__T10JsonParserTaVbi0Z10JsonParser5StateVii16Z.Stack.error(ulong) [0xf2f4bc]
./src/ocean/util/container/more/Stack.d:169 pure nothrow @safe ocean.text.json.JsonParser.JsonParser!(char, false).JsonParser.State ocean.util.container.more.Stack.__T5StackTE5ocean4text4json10JsonParser22__T10JsonParserTaVbi0Z10JsonParser5StateVii16Z.Stack.pop() [0xf2f0ae]
./src/ocean/text/json/JsonParser.d:418 pure nothrow bool ocean.text.json.JsonParser.JsonParser!(char, false).JsonParser.pop(ocean.text.json.JsonParser.JsonParser!(char, false).JsonParser.Token) [0xbea764]
./src/ocean/text/json/JsonParser.d:224 bool ocean.text.json.JsonParser.JsonParser!(char, false).JsonParser.parseMemberName() [0xbea19b]
./src/ocean/text/json/JsonParser.d:122 bool ocean.text.json.JsonParser.JsonParser!(char, false).JsonParser.next() [0xbe9cb4]
./src/ocean/text/json/JsonParser.d:686 void ocean.text.json.JsonParser.__unittestL663_3() [0xbe99e4]
??:? void ocean.text.json.JsonParser.__modtest() [0xbec575]
./src/ocean/core/UnitTestRunner.d:648 ocean.core.UnitTestRunner.UnitTestRunner.Result ocean.core.UnitTestRunner.UnitTestRunner.timedTest(object.ModuleInfo*, out core.sys.posix.sys.time.timeval, ref char[]) [0xe9dea6]
./src/ocean/core/UnitTestRunner.d:265 _D5ocean4core14UnitTestRunner14UnitTestRunner3runMFZ14__foreachbody1MFPS6object10ModuleInfoZi [0xe9ce69]
src/object.d:1585 int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)).__lambda2(immutable(object.ModuleInfo*)) [0x1060e97]
src/rt/minfo.d:289 int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))).__foreachbody2(ref rt.sections_elf_shared.DSO) [0x106ab4e]
src/rt/sections_elf_shared.d:53 int rt.sections_elf_shared.DSO.opApply(scope int delegate(ref rt.sections_elf_shared.DSO)) [0x106b019]
src/rt/minfo.d:282 int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))) [0x106aadf]
src/object.d:1585 int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)) [0x1060e73]
./src/ocean/core/UnitTestRunner.d:209 int ocean.core.UnitTestRunner.UnitTestRunner.run() [0xe9c9af]
./src/ocean/core/UnitTestRunner.d:943 _Dmain [0xe9e94f]
src/rt/dmain2.d:472 _D2rt6dmain211_d_run_mainUiPPaPUAAaZiZ6runAllMFZ9__lambda1MFZv [0x1066226]
src/rt/dmain2.d:447 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x1066164]
src/rt/dmain2.d:472 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).runAll() [0x10661e2]
src/rt/dmain2.d:447 void rt.dmain2._d_run_main(int, char**, extern (C) int function(char[][])*).tryExec(scope void delegate()) [0x1066164]
src/rt/dmain2.d:480 _d_run_main [0x10660c1]
??:? main [0xe9ed67]
??:? __libc_start_main [0xb7b9d82f]
submodules/makd/Makd.mak:513: recipe for target '/home/daniz/projects/ocean-1/build/devel/tmp/allunittests.stamp' failed
make: *** [/home/daniz/projects/ocean-1/build/devel/tmp/allunittests.stamp] Error 1
```

--- a/src/ocean/util/container/more/Stack.d
+++ b/src/ocean/util/container/more/Stack.d
@@ -17,7 +17,9 @@
 
 module ocean.util.container.more.Stack;
 
-import ocean.core.ExceptionDefinitions : ArrayBoundsException;
+import ocean.core.Enforce;
+version (UnitTest)
+    import ocean.core.Test;
 
 /******************************************************************************
 
@@ -133,7 +135,7 @@ struct Stack (V, int Size = 0)
                           if (depth < stack.length)
                               stack[depth++] = value;
                           else
-                             error (__LINE__);
+                              enforce(.e_bounds, false);
                           }
                 return this;
         }
@@ -166,7 +168,8 @@ struct Stack (V, int Size = 0)
                 if (depth)
                     return stack[--depth];
 
-                return error (__LINE__);
+                enforce(.e_bounds, false);
+                assert(false);
         }
 
         /**********************************************************************
@@ -182,7 +185,8 @@ struct Stack (V, int Size = 0)
                 if (depth)
                     return stack[depth-1];
 
-                return error (__LINE__);
+                enforce(.e_bounds, false);
+                assert(false);
         }
 
         /**********************************************************************
@@ -203,7 +207,8 @@ struct Stack (V, int Size = 0)
                    return p[1] = v;
                    }
 
-                return error (__LINE__);
+                enforce(.e_bounds, false);
+                assert(false);
         }
 
         /**********************************************************************
@@ -220,7 +225,8 @@ struct Stack (V, int Size = 0)
                 if (i < depth)
                     return stack [depth-i-1];
 
-                return error (__LINE__);
+                enforce(.e_bounds, false);
+                assert(false);
         }
 
         /**********************************************************************
@@ -242,7 +248,7 @@ struct Stack (V, int Size = 0)
                    *p = t;
                    }
                 else
-                   error (__LINE__);
+                   enforce(.e_bounds, false);
                 return this;
         }
 
@@ -265,7 +271,7 @@ struct Stack (V, int Size = 0)
                    *p = t;
                    }
                 else
-                   error (__LINE__);
+                   enforce(.e_bounds, false);
                 return this;
         }
 
@@ -284,17 +290,6 @@ struct Stack (V, int Size = 0)
                 return stack [0 .. depth];
         }
 
-        /**********************************************************************
-
-                Throw an exception
-
-        **********************************************************************/
-
-        private V error (size_t line)
-        {
-                throw new ArrayBoundsException (__FILE__, line);
-        }
-
         /***********************************************************************
 
                 Iterate from the most recent to the oldest stack entries
@@ -311,6 +306,49 @@ struct Stack (V, int Size = 0)
         }
 }
 
+///
+unittest
+{
+    Stack!(int) stack;
+    stack.push(42);
+    test!("==")(stack.pop(), 42);
+    testThrown!(StackBoundsException)(stack.pop());
+}
+
+/*******************************************************************************
+
+    Exception that indicates any kind of out of bound access in stack, for
+    example, trying to pop from empty one.
+
+*******************************************************************************/
+
+public class StackBoundsException : ExceptionBase
+{
+    this ( )
+    {
+        version (D_Version2)
+            super("Out of bounds access attempt to stack struct");
+        else
+            super("", 0);
+    }
+}
+
+// HACK: some D1 code may be trying to catch ArrayBoundsException specifically
+// so different bases are used for smoother migration.
+version (D_Version2)
+    private alias Exception ExceptionBase;
+else
+{
+    import ocean.core.ExceptionDefinitions : ArrayBoundsException;
+    private alias ArrayBoundsException ExceptionBase;
+}
+
+private StackBoundsException e_bounds;
+
+static this ( )
+{
+    e_bounds = new StackBoundsException;
+}
 
 /*******************************************************************************
 

--- a/src/ocean/util/container/more/Stack.d
+++ b/src/ocean/util/container/more/Stack.d
@@ -244,7 +244,10 @@ struct Stack (V, int Size = 0)
                    auto p = &stack[depth-d];
                    auto t = *p;
                    while (--d)
-                          *p++ = *(p+1);
+                      {
+                          *p = *(p+1);
+                          p++;
+                      }
                    *p = t;
                    }
                 else
@@ -267,7 +270,10 @@ struct Stack (V, int Size = 0)
                    auto p = &stack[depth-1];
                    auto t = *p;
                    while (--d)
-                          *p-- = *(p-1);
+                      {
+                          *p = *(p-1);
+                          p--;
+                      }
                    *p = t;
                    }
                 else

--- a/src/ocean/util/container/more/Stack.d
+++ b/src/ocean/util/container/more/Stack.d
@@ -321,6 +321,77 @@ unittest
     testThrown!(StackBoundsException)(stack.pop());
 }
 
+version(UnitTest)
+{
+    static void runTests ( T ) ( NamedTest t, T stack )
+    {
+        t.test!("==")(stack.size(), 0);
+        testThrown!(StackBoundsException)(stack.pop());
+        stack.push(42);
+        t.test!("==")(stack.size(), 1);
+        stack.clear();
+        t.test!("==")(stack.size(), 0);
+
+        stack.push(100);
+        t.test!("==")(stack.dup(), 100);
+        t.test!("==")(stack[], [ 100, 100 ]);
+
+        auto clone = stack.clone();
+        foreach (idx, ref field; clone.tupleof)
+            t.test!("==")(field, stack.tupleof[idx]);
+
+        stack.clear();
+        stack.append(1, 2, 3, 4);
+        t.test!("==")(stack[], [ 1, 2, 3, 4 ]);
+
+        t.test!("==")(stack.top(), 4);
+        t.test!("==")(stack.pop(), 4);
+        t.test!("==")(stack.top(), 3);
+        t.test!("==")(stack[0], 3);
+        t.test!("==")(stack[2], 1);
+        testThrown!(StackBoundsException)(stack[10]);
+
+        stack.swap();
+        t.test!("==")(stack[], [ 1, 3, 2 ]);
+        stack.clear();
+        testThrown!(StackBoundsException)(stack.swap());
+
+        stack.append(1, 2, 3);
+        stack.rotateLeft(2);
+        t.test!("==")(stack[], [ 1, 3, 2 ]);
+        stack.rotateRight(2);
+        t.test!("==")(stack[], [ 1, 2, 3 ]);
+        testThrown!(StackBoundsException)(stack.rotateLeft(40));
+        testThrown!(StackBoundsException)(stack.rotateRight(40));
+    }
+}
+
+unittest
+{
+    // common tests
+    runTests(new NamedTest("Dynamic size"), Stack!(int).init);
+    runTests(new NamedTest("Static size"),  Stack!(int, 10).init);
+}
+
+unittest
+{
+    // fixed size specific tests
+    Stack!(int, 3) stack;
+    test!("==")(stack.unused(), 3);
+
+    stack.push(1);
+    stack.push(1);
+    stack.push(1);
+    testThrown!(StackBoundsException)(stack.push(1));
+}
+
+unittest
+{
+    // dynamic size specific tests
+    Stack!(int) stack;
+    test!("==")(stack.unused(), 0);
+}
+
 /*******************************************************************************
 
     Exception that indicates any kind of out of bound access in stack, for

--- a/src/ocean/util/container/more/Stack.d
+++ b/src/ocean/util/container/more/Stack.d
@@ -93,7 +93,7 @@ struct Stack (V, int Size = 0)
 
         Stack clone ()
         {
-                Stack s = void;
+                Stack s;
                 static if (Size == 0)
                            s.stack.length = stack.length;
                 s.stack[] = stack;

